pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
 
 	return csr, nil
 }
+
+type PEMBundle struct {
+	CAPEM    []byte
+	ChainPEM []byte
+}
+
+type chainNode struct {
+	cert *x509.Certificate
+
+	issuer *chainNode
+}
+
+// ParseCertificateChainPEM will decode the bundle before passing to
+// ParseCertificateChain
+func ParseCertificateChainPEM(pembundle []byte) (*PEMBundle, error) {
+	certs, err := DecodeX509CertificateChainBytes(pembundle)
+	if err != nil {
+		return nil, err
+	}
+	return ParseCertificateChain(certs)
+}
+
+// ParseCertificateChain will return an x509 PEM encoded certificate chain and
+// CA. The CA may not be a true root, but the highest intermediate certificate.
+// Removes duplicate entries. The returned CA may be empty if a single
+// certificate was passed. Chain contains leaf certificate first. Removes
+// comments and white space. An error is returned if the passed bundle is not a
+// valid flat tree chain, the bundle is malformed, or the chain is broken.
+func ParseCertificateChain(certs []*x509.Certificate) (*PEMBundle, error) {
+	// De-duplicate certificates
```

Comment:
Thanks, that makes sense!

That said, RFC 8443 actually mentions that TLS 1.3 clients should be able to deal with duplication. Isn't deduplicating in very-manager going to hide the dupe issue from operations people, since we know that clients are already able to deal with duplication?

Would it be possible to add this information to the comment that says "Deduplicate certificates"?

Reply:
Hmm, just because clients _can_ deal with duplication, doesn't mean we shouldn't bother in cert-manager. It makes sense to me to write code that actively cleans up these kinds of ugliness that other issuers present to us.

Regardless, allowing duplicates breaks this code which is expecting a single flat chain, where every item is a child for i+1, and a parent for i-1. Allowing for duplicates doesn't make much sense to me.

Reply:
> Hmm, just because clients can deal with duplication, doesn't mean we shouldn't bother in cert-manager. It makes sense to me to write code that actively cleans up these kinds of ugliness that other issuers present to us.

Strong agree on this; especially where we're dealing with RSA certs, cleaning up duplicates will save a _huge_ chunk of the size of the TLS handshake.

I _can_ imagine situations - as the TLS 1.3 spec snippet Mael posted hints at - where a server might intentionally want to send two intermediates in one chain as some clients will accept that, but I think that such scenarios would be really rare.